unified-runtime/source/adapters/level_zero/event.cpp

```diff
@@ -884,12 +887,20 @@ ur_result_t
 urEventRelease(/** [in] handle of the event object */ ur_event_handle_t Event) {
   Event->RefCountExternal--;
   bool isEventsWaitCompleted =
-      Event->CommandType == UR_COMMAND_EVENTS_WAIT && Event->Completed;
+      (Event->CommandType == UR_COMMAND_EVENTS_WAIT ||
+       Event->CommandType == UR_COMMAND_EVENTS_WAIT_WITH_BARRIER) &&
+      Event->Completed;
   UR_CALL(urEventReleaseInternal(Event));
   // If this is a Completed Event Wait Out Event, then we need to cleanup the
   // event at user release and not at the time of completion.
   if (isEventsWaitCompleted) {
+    auto EventRefcnt = Event->RefCount.load();
```

Comment:
This race'y. You can't read and act on this variable, since it can immediately change from another thread. If this thread has an event handle, the refcount must be at greater than 0.

Reply:
So, to avoid the memory leak, we have to check if the CleanupCompletedEvent actually released the event. When this still leaks the event is when the event is already marked as "completed" and we don't decrement/release the event.

The alternative is to pass a flag to CleanupCompletedEvent to determine if the event was released or not.

Reply:
Actually, we have a flag that could be checked instead, Event->CleanedUp , let me add that check

Reply:
Fix in the latest update